File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/shape/Triangle2D.java
package com.greentree.commons.geometry.geom2d.shape;

import com.greentree.commons.geometry.geom2d.util.Transform2DImpl;
import com.greentree.commons.math.Mathf;
import com.greentree.commons.math.vector.AbstractVector2f;
import com.greentree.commons.math.vector.Vector2f;

public final class Triangle2D extends Shape2D {

    public Triangle2D(AbstractVector2f v1, AbstractVector2f v2, AbstractVector2f v3) {
        super(new Transform2DImpl(), points(v1, v2, v3));
    }

    private static AbstractVector2f[] points(AbstractVector2f v1, AbstractVector2f v2,
                                             AbstractVector2f v3) {
        return new AbstractVector2f[]{v1, v2, v3};
    }

    public Triangle2D(float x1, float y1, float x2, float y2, float x3, float y3) {
        super(new Transform2DImpl(), points(x1, y1, x2, y2, x3, y3));
    }

    private static AbstractVector2f[] points(float x1, float y1, float x2, float y2, float x3,
                                             float y3) {
        return points(vec(x1, y1), vec(x2, y2), vec(x3, y3));
    }

    private static AbstractVector2f vec(float x, float y) {
        return new Vector2f(x, y);
    }

    @Override
    public Triangle2D clone() {
        throw new UnsupportedOperationException();
    }

    @Override
    public float getArea() {
        var v1 = p2().minus(p1());
        var v2 = p3().minus(p1());
        return Mathf.abs(v1.cross(v2)) / 2f;
    }

    public AbstractVector2f p2() {
        return getPoints()[1];
    }

    public AbstractVector2f p1() {
        return getPoints()[0];
    }

    public AbstractVector2f p3() {
        return getPoints()[2];
    }

}

File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/shape/Triangle2D.kt
package com.greentree.commons.geometry.geom2d.shape

import com.greentree.commons.math.vec2f
import org.joml.Vector2fc

data class Triangle2D(
	val p1: Vector2fc,
	val p2: Vector2fc,
	val p3: Vector2fc,
) : FiniteShape2D {

	override val points = arrayOf(p1, p2, p3)

	constructor(
		x1: Float, y1: Float,
		x2: Float, y2: Float,
		x3: Float, y3: Float,
	) : this(
		vec2f(x1, y1),
		vec2f(x2, y2),
		vec2f(x3, y3)
	)
}
